Add export_panorama_attributes_to_csv

Avery-H opened the pull request:
Just like get_panorama_attributes which prints the
panorama attributes in json to standard out, this
method prints them out to multiple csv files. It will
create a new folder (named based on the pto file
input), 3 csv files, and an XLSX file with all 3 csv
files as individual worksheets.

In order to create this XLSX file I created a new module
called XLSX_Writer which uses the write_xlsx gem. Because
this gem only supports writing new XLSX files we have to
write the XLSX file with all 3 csv files within the same
method. Saying that the csv_to_xlsx method takes either a
string with a singular csv file name, or an array with
multiple file names.

https://www.wrike.com/open.htm?id=566725406

stowersjoshua commented on lib/panomosity/xlsx_writer.rb:
```diff
@@ -0,0 +1,51 @@
+      logger.info "Done. Check for #{output_file}.xlsx"
+    end
+  end
+end
```
Missing null termination.

stowersjoshua commented on lib/panomosity/xlsx_writer.rb:
```diff
@@ -0,0 +1,51 @@
+
+        File.open(input_files, 'r') do |f|
+          f.each_line do |rows|
+            cells = rows.split(/,(?![^\[]*\])/)
```
Regex should be assigned to a well named variable to help us understand what it is doing.

stowersjoshua replied:
Since it's the same as line 18, it should probably be a static variable.

stowersjoshua commented on lib/panomosity/xlsx_writer.rb:
```diff
@@ -0,0 +1,51 @@
+require 'write_xlsx'
+
+module Panomosity
+  module XLSX_Writer
```
RSGesus says not do use underscores in class or module names.

https://github.com/rubocop-hq/ruby-style-guide#camelcase-for-classes

It's example is:
```ruby
class XMLSomething
end
```

stowersjoshua replied:
If you want to be cool this could be a class instead so you can use instance variables and abstract out logic without cluttering other classes with unusable methods.

```ruby
class XLSXWriter
  attr_accessor :input_files, :output_file, :workbook

  def initialize(*input_files, output:)
    @input_files = input_files
    @output_file = output_file
    @workbook = WriteXLSX.new("#{output_file}.xlsx")
  end

  def save
    input_files.each do |input_file|
      write_worksheet(input_file|)
    end

    workbook.close
  end

  private

  def write_worksheet(input_file)
    workbook.add_worksheet(input_file
    # etc.
  end
end 

writer = XLSXWriter.new('aaaa.csv', 'bbbb.csv', output: 'aabb')
writer.save
```

stowersjoshua commented on lib/panomosity/runner.rb:
```diff
@@ -490,6 +493,70 @@ def get_panorama_attributes
+
+      csv_to_xlsx(['control_points.csv', 'cp_pairs.csv', 'neighborhood_pairs.csv'], 'attributes')
+
+      Dir.chdir '..'
```
@evanscottgray

evanscottgray replied:
Let’s use --output-dir or something to let users specify which directory to put it in, then name consider naming the files with the pto file name as a prefix instead.

Ideally, users should also be able to specify if they want csv, or excel, and they should also be able to select which files they want, for example, what if only the control point file is desired?

Additionally, we should consider adding an option to allow printing to stdout if it is a csv.

stowersjoshua replied:
@Avery-H promises to handle this in a follow-up PR

evanscottgray replied:
👍

Avery-H replied:
What is this javascript?

stowersjoshua replied:
[Yes](https://opalrb.com/)

stowersjoshua commented on lib/panomosity/runner.rb:
```diff
@@ -490,6 +493,70 @@ def get_panorama_attributes
+      logger.info "Making control_points.csv"
+
+      filename = 'control_points.csv'
+      headers = %w(image_1_name image_2_name image_1_id image_2_id type width height d1 e1 d2 e2 x1 y1 x2 y2 rx ry r)
```
We use hard brackets for our literal arrays

https://github.com/rubocop-hq/ruby-style-guide#percent-literal-braces

stowersjoshua commented on lib/panomosity/xlsx_writer.rb:
```diff
@@ -0,0 +1,51 @@
+          end
+        end
+      else
+        logger.info "Creating #{input_files} Worksheet"
```
This will not log cleanly:
```ruby
> puts "hello #{%w[aaa bbb ccc]} there"
# => hello ["aaa", "bbb", "ccc"] there
```

Maybe use something like:
```
logger.info "Creating Worksheet from #{input_files.join(', ')}"
```

Avery-H replied:
This will always be a string

stowersjoshua commented on lib/panomosity/xlsx_writer.rb:
```diff
@@ -0,0 +1,51 @@
+
+module Panomosity
+  module XLSX_Writer
+    def csv_to_xlsx(input_files, output_file)
```
Instead of handling single and multi file inputs separately, why not make all single file inputs into arrays?

```ruby
def csv_to_xlsx(input_files, output_file)
  input_files = Array(input_files)
  workbook = WriteXLSX.new("#{output_file}.xlsx")

  input_files.each do |csv|
    worksheet = workbook.add_worksheet(csv)

    File.open(csv, 'r') do |input_file|
      input_file.each_line do |rows|
        magical_regex = /,(?![^\[]*\])/
        cells = rows.split(magical_regex)

        cells.each_with_index do |cell, column|
          row = input_file.lineno - 1
          data = cell.tr_s('"', '').strip
          worksheet.write(row, column, data)
        end
      end
    end
  end

  workbook.close
end
```

Avery-H replied:
This is actually how I originally wrote it. I think I changed it to take both a string or an array simple because an array with one string felt wrong to me.

stowersjoshua replied:
If the files were handled differently based on if there is one or many, I would totally agree.
But since we do the same thing to it either way, this makes it more difficult to maintain instead.

If I come along later and want to change the writing logic, I would have to know to change it in both places. This design would make me think that one is supposed to be different than the other.

stowersjoshua commented on lib/panomosity/xlsx_writer.rb:
```diff
@@ -0,0 +1,51 @@
+
+      workbook = WriteXLSX.new("#{output_file}.xlsx")
+
+      if input_files.is_a? Array
```
In this repo, we use parentheses here.

https://github.com/elevatesystems/panomosity/pull/4#discussion_r497223367

Avery-H replied:
#ThatGuyJosh

stowersjoshua reviewed (changes requested):
Great abstraction!

evanscottgray reviewed (changes requested):
Overall looking good, have a few more comments but on mobile right now, will add asap.

Avery-H committed:
Add export_panorama_attributes_to_csv

Just like get_panorama_attributes which prints the
panorama attributes in json to standard out, this
method prints them out to multiple csv files. It will
create a new folder (named based on the pto file
input), 3 csv files, and an XLSX file with all 3 csv
files as individual worksheets.

In order to create this XLSX file I created a new module
called XLSX_Writer which uses the write_xlsx gem. Because
this gem only supports writing new XLSX files we have to
write the XLSX file with all 3 csv files within the same
method. Saying that the csv_to_xlsx method takes either a
string with a singular csv file name, or an array with
multiple file names.

Signed-off-by: Avery Hankins <averyhankinstech@gmail.com>

Avery-H committed:
Bump Version number

Include gemspec and Gemfile.lock

Signed-off-by: Avery Hankins <averyhankinstech@gmail.com>
